### PR TITLE
Add linter configuration

### PR DESCRIPTION
**Pull request by epk**
## Description

Adding a linter configuration file for deterministic linting

### CODE REVIEW

**Comment by epk** [comment]
Sorry about the excessive review pings

File: .golangci.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  enable:
+    # Defaults
+    - deadcode
+    - gosimple
+    - govet
+    - ineffassign
+    - staticcheck
+    - structcheck
+    - typecheck
+    - varcheck
```
Comment (shlomi-noach) [inline .golangci.yml:19]: How did you pick this specific list of flags? (To clarify, I'm not sure what the list should be)

File: .golangci.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+    # Extras
+    - gofmt
+    - goimports
+
+# https://github.com/golangci/golangci/wiki/Configuration
+service:
+  golangci-lint-version: 1.34.x # use the fixed version to not introduce new linters unexpectedly
```
Comment (shlomi-noach) [inline .golangci.yml:27]: +1 for fixed version

File: misc/git/hooks/golangci-lint [hunk 3]
```diff
@@ -25,12 +25,12 @@ gopackages=$(echo $gofiles | xargs -n1 dirname | sort -u)
 GOLANGCI_LINT=$(command -v golangci-lint >/dev/null 2>&1)
 if [ $? -eq 1 ]; then
   echo "Downloading golangci-lint..."
-  curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.27.0
+  curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.34.1
 fi
 
 #golangci-lint run --disable=ineffassign,unused,gosimple,staticcheck,errcheck,structcheck,varcheck,deadcode
```
Comment (shlomi-noach) [inline misc/git/hooks/golangci-lint:31]: Should we remove this commented line?

File: misc/git/hooks/golangci-lint [hunk 4]
```diff
@@ -25,12 +25,12 @@ gopackages=$(echo $gofiles | xargs -n1 dirname | sort -u)
...
 for gopackage in $gopackages
 do
  echo "Linting $gopackage"
- golangci-lint run --disable=errcheck --timeout=10m $gopackage
+ golangci-lint run $gopackage
```
Comment (shlomi-noach) [inline misc/git/hooks/golangci-lint:35]: Since `errcheck` is removed, should we expect new linter errors on existing code?

**Review by shlomi-noach — COMMENTED** [review]
Thanks so much for working on this, please see questions inline

File: misc/git/hooks/golangci-lint (hunk 4 above)
Comment (ajm188) [reply misc/git/hooks/golangci-lint:35]: `errcheck` is implicitly removed by the combination of `disable-all: true` and `errcheck` not being in the enable list in the config file

Comment (ajm188) [reply misc/git/hooks/golangci-lint:35]: Also, I wanted to ask: for the commit hook specifically, can we add the `--fast` flag?

File: .golangci.yml (hunk 1 above)
Comment (epk) [reply .golangci.yml:19]: I ran `golangci-lint linters` in a directory without a `.golangci.yml` present

```
 ~  golangci-lint linters
Enabled by your configuration linters:
deadcode: Finds unused code [fast: true, auto-fix: false]
errcheck: Errcheck is a program for checking for unchecked errors in go programs. These unchecked errors can be critical bugs in some cases [fast: true, auto-fix: false]
gosimple (megacheck): Linter for Go source code that specializes in simplifying a code [fast: true, auto-fix: false]
govet (vet, vetshadow): Vet examines Go source code and reports suspicious constructs, such as Printf calls whose arguments do not align with the format string [fast: true, auto-fix: false]
ineffassign: Detects when assignments to existing variables are not used [fast: true, auto-fix: false]
staticcheck (megacheck): Staticcheck is a go vet on steroids, applying a ton of static analysis checks [fast: true, auto-fix: false]
structcheck: Finds unused struct fields [fast: true, auto-fix: false]
typecheck: Like the front-end of a Go compiler, parses and type-checks Go code [fast: true, auto-fix: false]
unused (megacheck): Checks Go code for unused constants, variables, functions and types [fast: false, auto-fix: false]
varcheck: Finds unused global variables and constants [fast: true, auto-fix: false]
```

I only removed `errcheck` from the list as the previous CLI invocations had it disabled

**Review by ajm188 — COMMENTED** [review]
config looks good to me. looks like vtctl.go isn't goimports'ed yet

**Comment by epk** [comment]
> config looks good to me. looks like vtctl.go isn't goimports'ed yet

I don't know why it is failing in the CI, running the linter locally it passes.

It also passed on a previous commit with no changes :/

**Comment by shlomi-noach** [comment]
> I don't know why it is failing in the CI, running the linter locally it passes.

Also passes for me, but I notice:
```
$ $(go env GOPATH)/bin/golangci-lint --version
golangci-lint has version 1.31.0 built from 3d6d0e7 on 2020-09-07T15:14:41Z
```

so I'm using `1.31.0`, and CI presumably uses `1.34`? @epk Can you verify what your own local `golangci-lint` version is?

**Comment by epk** [comment]
> > I don't know why it is failing in the CI, running the linter locally it passes.
> 
> Also passes for me, but I notice:
> 
> ```
> $ $(go env GOPATH)/bin/golangci-lint --version
> golangci-lint has version 1.31.0 built from 3d6d0e7 on 2020-09-07T15:14:41Z
> ```
> 
> so I'm using `1.31.0`, and CI presumably uses `1.34`? @epk Can you verify what your own local `golangci-lint` version is?

I have the same version (1.34.1) locally:

```
~/src/vitess.io/vitess   epk/golangci-lint  $(go env GOPATH)/bin/golangci-lint --version
golangci-lint has version 1.34.1 built from d7dd233 on 2020-12-29T06:20:54Z
~/src/vitess.io/vitess   epk/golangci-lint  $(go env GOPATH)/bin/golangci-lint run go/...
~/src/vitess.io/vitess   epk/golangci-lint  echo $?
0
```

**Comment by shlomi-noach** [comment]
> I have the same version (1.34.1) locally:

and, on the server side?

**Comment by shlomi-noach** [comment]
Have resovled merge conflicts and am further removing linter errors

**Comment by shlomi-noach** [comment]
Pinned `golang-lint` version to `1.31.0`

**Comment by shlomi-noach** [comment]
Linter tests now passing

**Review by shlomi-noach — APPROVED** [review]

**Pull request merged**